### PR TITLE
materialized: add local process orchestrator

### PR DESCRIPTION
**Pull request by benesch**
Add a new `Orchestrator` backend, `ProcessOrchestrator` that manages
local processes. This sits alongside the existing
`KubernetesOrchestrator` backend.

The `ProcessOrchestrator` is intended for local development. It is much
more convenient to have `materialized` spin up `dataflowd` processes as
necessary than to have to futz with Minikube.

The big defect with this orchestrator is that you have to remember to
build `dataflowd`. If you are not careful, you can end up with
`materialized` and `dataflowd` processes that are out of sync. In other
words, it's a bit too easy to accidentally test version compatibility
while doing local development with this orchestrator. We can fix this
later with e.g. a shell script that builds both dataflowd and
materialized.

To use the process orchestrator, simply run:

    $ cargo build --bin dataflowd
    $ cargo run -- --dev --orchestrator=process

The `materialized` process will automatically launch `dataflowd
--runtime=storage`. You can type

    # CREATE CLUSTER ... SIZE 'ignored'

to spin up separate `dataflowd --runtime=compute` processes.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Go commit by commit.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Review by antiguru — APPROVED** [review]
I think this looks good, but raises one question: Why do we still have the `dataflowd` binary if the `materialized` compilation unit probably contains everything that `dataflowd` includes. I'm tempted to remove `dataflowd` and integrate the functionality within `materialized`, guarded with the right flags. WDYT?

**Comment by philip-stoev** [comment]
This still allows parent processes to launch each type of Mz process independently , right? This is important for testing in general as the parent shoud know the pids of its children and be able to control them independently.

**Comment by benesch** [comment]
> This still allows parent processes to launch each type of Mz process independently , right? This is important for testing in general as the parent shoud know the pids of its children and be able to control them independently.

Yep! You still have the option of starting with no orchestrator and manually turning on `dataflowd` processes.

File: src/dataflow-types/src/client/controller.rs [hunk 1]
```diff
@@ -153,12 +156,7 @@ where
...
-                let addrs: Vec<_> = service
-                    .hosts()
-                    .iter()
-                    .map(|h| format!("{h}:2100"))
-                    .collect();
-                let client = RemoteClient::new(&addrs);
+                let client = RemoteClient::new(&service.addresses("controller"));
                 let client: Box<dyn ComputeClient<T>> = Box::new(client);
                 self.compute_mut(instance)
                     .unwrap()
```
Comment (antiguru) [inline src/dataflow-types/src/client/controller.rs:162]: We probably want to give the instance a better name than `"default"`, but we might not have the infrastructure in place to generate better names.

**Pull request merged**